### PR TITLE
implement person_data filter

### PR DESCRIPTION
**Pull request by nehahirve**
Related to #231

### CODE REVIEW

File: src/locale/misc/smartSearch/en.yml [hunk 1]
```diff
@@ -11,6 +11,7 @@ filterTitles:
     all: All people
     campaign_participation: People who have participated in a campaign
     most_active: A selection of the most active people
+    person_data: People who match some criteria
```
Comment (richardolsson) [inline src/locale/misc/smartSearch/en.yml:14]: This label is true for all the filters. Maybe replace it with something like "Search based on name, e-mail or similar fields"?

File: src/locale/misc/smartSearch/en.yml [hunk 2]
```diff
@@ -47,6 +48,29 @@ most_active:
...
+        first_name: first name
+        last_name: last name
+        gender: gender
+        co_address: co-address
+        city: city
+        zip_code: zip code
+        street_address: street address
+        alt_phone: alternative phone number
+        phone: phone number
+        remove: Remove this criteria
```
Comment (richardolsson) [inline src/locale/misc/smartSearch/en.yml:73]: Is this correct English? This/these criteria/criterion?

File: src/components/smartSearch/filters/PersonData/index.tsx [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    const ALL_FIELDS = Object.values(DATA_FIELD);
+
+    const initialCriteria = ALL_FIELDS
+        .filter(f => f in filter.config.fields)
+        .map(f => ({ field: f, value: filter.config.fields[f] || '' }));
+
+    const [criteria, setCriteria] = useState<Criterion[]>(initialCriteria);
+
+    const isSubmittable = criteria.every(c => c.value.length);
```
Comment (richardolsson) [inline src/components/smartSearch/filters/PersonData/index.tsx:35]: Nice! Logic is great, maybe some comments to explain?

File: src/components/smartSearch/filters/PersonData/index.tsx [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+const PersonData = ({ onSubmit, onCancel, filter: initialFilter }: PersonDataProps): JSX.Element => {
+    const { filter, setConfig, setOp } = useSmartSearchFilter<PersonDataFilterConfig>(initialFilter, { fields: { first_name: '' } });
+
+    const ALL_FIELDS = Object.values(DATA_FIELD);
+
+    const initialCriteria = ALL_FIELDS
+        .filter(f => f in filter.config.fields)
+        .map(f => ({ field: f, value: filter.config.fields[f] || '' }));
+
+    const [criteria, setCriteria] = useState<Criterion[]>(initialCriteria);
```
Comment (richardolsson) [inline src/components/smartSearch/filters/PersonData/index.tsx:33]: It can be quite difficult to visualize this structure, so I would recommend adding a comment that explains it.

File: src/components/smartSearch/filters/PersonData/index.tsx [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+    const handleValueChange = (value: string, c: Criterion) => {
+        setCriteria(criteria.map(criterion => {
+            return criterion.field === c.field ?
+                { ...criterion, value } : criterion;
+        }));
+    };
+
+    const handleSelectChange = (field: string, c: Criterion) => {
+        if (field === DEFAULT_VALUE) {
```
Comment (richardolsson) [inline src/components/smartSearch/filters/PersonData/index.tsx:67]: It was not clear to me when reading this that `DEFAULT_VALUE` was the option that is intended to delete a criterion.

The following two things are true about this constant:

1. It's value is the one that is selected by default in the select box
2. It's value is the one that is represented to users as "remove"

These are two attributes of this constant that are equally true, and could both be reflected in it's name. When naming this constant, the name can draw from either one of these two facts, i.e. it can be something like `DEFAULT_VALUE` or something like `REMOVE_OPTION`. Both of these will be obvious in one context and less obvious in the other:

1. Using a name like `DEFAULT_VALUE` raises no eyebrows when it's passed as the default value to a select box, but is less obvious in this context right here
2. Using a name like `REMOVE_OPTION` makes this use right here very clear, but is less inline with passing it as a default value

In lieu of using a name like `DEFAULT_VALUE_THAT_IS_ALSO_REMOVE_OPTION`, this is a necessary tradeoff to make. The name will either make obvious sense in one context, or in another, but likely not both.

So when picking a name, I think it's wise to go with the one which will add info/make sense in the context that is most difficult to read otherwise. In this case, I would argue that the more difficult-to-read context is the one right here, i.e. where an option is removed, and that the constant should thus be renamed to something like `REMOVE_OPTION` (not married to that specific name of course).

Passing the `REMOVE_OPTION` as the _de facto default value_ to the `value` property of the `StyledSelect` will very clearly communicate to a reader that the _remove_ option is the _default_ one.

More so than checking for `DEFAULT_VALUE` before doing some sort of filtering operation (which at closer glance turns out the be a delete). Why should picking the default value result in deleting?

So, bottom line is this: I think you should either rename this constant, or add some comments here that explains why it's used the way it is here, or both. 😊

File: src/components/smartSearch/filters/PersonData/index.tsx [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+                <Msg id="misc.smartSearch.person_data.addRemoveSelect.sub" />
+            </MenuItem>
+        </StyledSelect>
+    );
+
+    const getCriteriaString = () => {
+        let existing;
+        let criteriaString: JSX.Element | null = null;
+        criteria.forEach(c => {
+            const options = ALL_FIELDS.filter(f => f === c.field || !criteria.map(c => c.field).includes(f));
```
Comment (richardolsson) [inline src/components/smartSearch/filters/PersonData/index.tsx:95]: Performance is not really a problem for us here, but code like this still itches a little IMO. This could be optimized in many ways.

Right now there are four loops happening on this line.

1. The `criteria.forEach()` outer loop in which this code is being executed
2. `ALL_FIELDS.filter()` is just a loop that executes a function every iteration
3. The function executes (not always, but almost always) `criteria.map()` which will loop over the entire `criteria` array
4. Finally, after 3 finishes, `includes(f)` will loop for <= N iterations until `f` is found or not found.

I started writing an example of what is actually happening here, but if it's not entirely clear how some of this looping is redundant, lets instead look at it in a debugger. Let me know if you want to do that together!

There are a number of ways that this could be optimized (some of which can be combined, others are incompatible):

* Break `criteria.map()` out of the `criteria.forEach()` loop so that it's only executed once.
* Use `criteria.find()` or `criteria.some()` instead of `criteria.map().includes()`
* Prepare a single list of globally available options that match the `!criteria.map(…).includes(…)` condition, i.e. the ones that are not used anywhere, and use it everywhere instead of creating a new one every time. Append `c.field` (the only that differ) on every iteration.

File: src/components/smartSearch/filters/PersonData/index.tsx [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                            { criteria.length < ALL_FIELDS.length && (
+                                <StyledSelect
+                                    onChange={ e => addCriteria(e.target.value) }
+                                    value={ DEFAULT_VALUE }>
+                                    <MenuItem
+                                        key={ DEFAULT_VALUE }
+                                        value={ DEFAULT_VALUE }>
+                                        <Msg id="misc.smartSearch.person_data.ellipsis"/>
+                                    </MenuItem>
+                                    { ALL_FIELDS.filter(f => !criteria.map(c => c.field).includes(f)).map(o => (
```
Comment (richardolsson) [inline src/components/smartSearch/filters/PersonData/index.tsx:159]: Here's that loop again. Could be simplified same way as above.

**Review by richardolsson — COMMENTED** [review]
Awesome work! I think this works well and most of my comments below (especially the two longer ones) are not something you necessarily have to fix, but that I added mostly as feedback for your learning. 😊

File: src/locale/misc/smartSearch/en.yml (hunk 2 above)
Comment (nehahirve) [reply src/locale/misc/smartSearch/en.yml:73]: Technically, criteria is plural and criterion is singular, but as a linguist I would argue -- sort of correct, yes -- 
https://www.merriam-webster.com/words-at-play/criteria-vs-criterion-singular-plural-grammar#:~:text=Criteria%20is%20typically%20a%20plural,used%20singulars%20datum%20and%20agendum.

File: src/components/smartSearch/filters/PersonData/index.tsx (hunk 5 above)
Comment (nehahirve) [reply src/components/smartSearch/filters/PersonData/index.tsx:67]: v. good point! when you put it this way it's hard to un-see

File: src/components/smartSearch/filters/PersonData/index.tsx (hunk 6 above)
Comment (nehahirve) [reply src/components/smartSearch/filters/PersonData/index.tsx:95]: I've optimised it such that the currently selected option will render as the first option, then the global options below it ('remove this field' is always first). It doesn't feel weird to me but if it comes up as strange to use later on, I will change it, have made a note in the overall filter to do list!
